### PR TITLE
Adjust raft transactions to be safer with get operations

### PR DESCRIPTION
**Pull request by raskchanky**
This has the latest OSS changes from the updates to https://github.com/hashicorp/vault-enterprise/pull/2945

### CODE REVIEW

File: physical/raft/fsm.go [hunk 1]
```diff
@@ -642,6 +653,17 @@ func (f *FSM) ApplyBatch(logs []*raft.Log) []interface{} {
 						err = b.Put([]byte(op.Key), op.Value)
 					case deleteOp:
 						err = b.Delete([]byte(op.Key))
+					case getOp:
+						val := b.Get([]byte(op.Key))
+						if len(val) > 0 {
```
Comment (ncabatoff) [inline physical/raft/fsm.go:658]: Don't we want to append an entrySlice even when the len is 0?

Comment (raskchanky) [reply physical/raft/fsm.go:658]: If I'm following you here, this is to make sure we have the same number of entries in the slice as we have get ops? And then I guess at the other end, we just verify the value is present before updating anything.

Comment (raskchanky) [reply physical/raft/fsm.go:661]: Fixed in https://github.com/hashicorp/vault/pull/17151/commits/a5a96152575cf331726bb5b50de29bb9fcb65712

File: physical/raft/raft.go [hunk 2]
```diff
@@ -1622,10 +1617,26 @@ func (b *RaftBackend) applyLog(ctx context.Context, command *LogData) error {
...
 	}
 
-	if resp, ok := resp.(*FSMApplyResponse); !ok || !resp.Success {
+	fsmar, ok := resp.(*FSMApplyResponse)
+	if !ok || !fsmar.Success {
 		return errors.New("could not apply data")
 	}
 
+	// populate command with our results
+	if fsmar.EntrySlice != nil {
```
Comment (ncabatoff) [inline physical/raft/raft.go:1626]: When would EntrySlice be nil?

Comment (raskchanky) [reply physical/raft/raft.go:1626]: It wouldn't be, based on our code now, but it doesn't hurt to have a check for it, does it?

Comment (ncabatoff) [reply physical/raft/raft.go:1626]: Perhaps, perhaps not.  When I see an if like that, I'm forced to ask: what should the else case be?  You're protecting us against obvious failures (panic on nil) but possibly hiding a problem that could be harder to debug.  And a person reading the code is forced to wonder the same thing I did, and possibly go digging in the code to understand the scenario that gave rise to the if.

Comment (ncabatoff) [reply physical/raft/raft.go:1626]: I guess what I'm saying is: if it should never be nil, and you want to program defensively, then handle the nil case by returning an error.

Comment (raskchanky) [reply physical/raft/raft.go]: Fixed in https://github.com/hashicorp/vault/pull/17151/commits/bb5e0bd9a14255e1685c4a9bef463d0dd1dc4fe1

File: physical/raft/raft.go [hunk 3]
```diff
@@ -1622,10 +1617,28 @@ func (b *RaftBackend) applyLog(ctx context.Context, command *LogData) error {
...
+		return errors.New("entries on FSM response were empty")
+	}
+
+	for i, logOp := range command.Operations {
+		if logOp.OpType == getOp {
+			fsmEntry := fsmar.EntrySlice[i]
+
+			// this should always be true because the entries in the slice were created in the same order as
+			// the command operations.
+			if logOp.Key == fsmEntry.Key && len(fsmEntry.Value) > 0 {
```
Comment (ncabatoff) [inline physical/raft/raft.go:1636]: I suppose if you're going to worry about a nil EntrySlice, you should also maybe return an error when when the key doesn't match.  Up to you.

**Review by ncabatoff — APPROVED** [review]

**Pull request merged**